libs/injection/bbhnet/injection/injection.py

```diff
@@ -0,0 +1,261 @@
+        signal = np.zeros(waveform_size)
+        b, a = sig.butter(
+            N=8,
+            Wn=waveform_generator.waveform_arguments["minimum_frequency"],
+            btype="highpass",
+            fs=waveform_generator.sampling_frequency,
+        )
+        for mode in polarizations.keys():
+            # Get ifo response
+            response = ifo.antenna_response(ra, dec, geocent_time, psi, mode)
+            polarizations[mode] = sig.filtfilt(b, a, polarizations[mode])
+            signal += polarizations[mode] * response
+
+        # Total shift = shift to trigger time + geometric shift
+        dt = waveform_duration / 2.0
+        dt += ifo.time_delay_from_geocenter(ra, dec, geocent_time)
+        signal = np.roll(signal, int(np.round(dt * sample_rate)))
+
+        # Calculate SNR
+        if noise_psd is not None:
+            if get_snr:
+                snr[i] = get_snr(signal, noise_psd, sample_rate)
+
+        signals[i] = signal
+    if get_snr:
+        return signals, snr
+    return signals
+
+
+def inject_signals(
+    frame_files: [str],
+    channels: [str],
+    ifos: [str],
+    prior_file: str,
+    n_samples: int,
+    outdir: str,
+    fmin: float = 20,
+    waveform_duration: float = 8,
+    snr_range: [float] = [25, 50],
+):
+
+    """ Start simulation """
+
+    strain = TimeSeries.read(frame_files[0], channels[0])
+    logging.info(f"Read strain from {frame_files[0]}")
+
+    frame_start, frame_stop = strain.span
+    frame_duration = frame_stop - frame_start
+
+    sample_rate = int(strain.sample_rate.value)
+    fftlength = int(max(2, np.ceil(2048 / sample_rate)))
+
+    # set the non-overlapping times of the signals in the frames randomly
+    # leaves buffer at either end of the series so edge effects aren't an issue
+    signal_times = sorted(
+        np.random.choice(
+            np.arange(
+                waveform_duration,
+                frame_duration - waveform_duration,
+                waveform_duration,
+            ),
+            size=n_samples,
+            replace=False,
+        )
+    )
+
+    # log and print out some simulation parameters
+    logging.info("Simulation parameters")
+    logging.info("Number of samples     : {}".format(n_samples))
+    logging.info("Sample rate [Hz]      : {}".format(sample_rate))
+    logging.info("High pass filter [Hz] : {}".format(fmin))
+    logging.info("Prior file            : {}".format(prior_file))
+
+    # define a Bilby waveform generator
+    waveform_generator = bilby.gw.WaveformGenerator(
+        duration=waveform_duration,
+        sampling_frequency=sample_rate,
+        frequency_domain_source_model=lal_binary_black_hole,
+        parameter_conversion=convert_to_lal_binary_black_hole_parameters,
+        waveform_arguments={
+            "waveform_approximant": "IMRPhenomPv2",
+            "reference_frequency": 50,
+            "minimum_frequency": 20,
+        },
+    )
```

Comment:
So the assumption here is that the sample rate of the first frame file is the same as the sample rate for the rest of the frame files, correct?

Would it maybe make sense to have `sample_rate` be an argument to the function, then resample the strain timeseries to this rate upon reading? Or is it worth writing files "as-is", with the only difference being the injection of strain?

Alternatively, is this a time consuming step? Could we instantiate a new waveform generator for each file?

Otherwise, if this is going to be the most efficient way of doing things, it's probably worthwhile to insert a check for each file in case it doesn't match, to make sure we're not generating bad data.

Reply:
Added check to ensure all frames have the same sample rate